shared/test/middleware/helpers/test_storage_apps.rb

```diff
@@ -98,4 +98,25 @@ def test_users_paired_on_level_with_level_paired
     # If there is a paired_user_level, then the level was paired.
     assert storage_apps.users_paired_on_level?(12345, 123, 124, 67890)
   end
+
+  def test_derive_project_type
+    signedin_storage_id = @user_storage_ids_table.insert(user_id: 20)
+    storage_apps = StorageApps.new(signedin_storage_id)
+
+    # Create without type
+    typeless = storage_apps.create({}, ip: 123)
+    assert_equal 'unknown', storage_apps.project_type_from_channel_id(typeless)
+
+    # Create with type argument
+    type_in_column = storage_apps.create({}, ip: 123, type: 'applab')
+    assert_equal 'applab', storage_apps.project_type_from_channel_id(type_in_column)
+
+    # Create with type in value JSON blob
+    type_in_json = storage_apps.create({projectType: 'gamelab'}, ip: 123)
+    assert_equal 'gamelab', storage_apps.project_type_from_channel_id(type_in_json)
+
+    # Create with type that can be derived from the level property in the JSON blob
+    type_in_level = storage_apps.create({level: '/projects/weblab'}, ip: 123)
+    assert_equal 'weblab', storage_apps.project_type_from_channel_id(type_in_level)
```

Comment:
Question: Should it be valid to create a project in all these different ways?  Seems like there's a lot of missing validation here.

Reply:
It would certainly be nice to reduce the number of ways that the project type is stored, and the number of ways you've uncovered that it can currently done certainly seems like a lot to keep in mind when writing new features like this. I'd like to prioritize fixing this after May/June deadlines.

When the "channels api" was created, it was deliberately made to be not specific to projects, and so it made sense for pixelation to use it without a project type. However, over time, we've added project-specific assumptions to the channels api and storage apps table, without explicitly changing the contract to say that a "channel" is now always a "project".

One path forward for pixelation levels is to make them look the same as hidden projects on other project-backed script levels, most notably by including a project_type in the db row.

Does that answer your question?

Reply:
Yep, that explains quite a bit.  I didn't totally understand channels was not supposed to be specific to projects.